docs/design/proposals/references/README.md

```diff
@@ -0,0 +1,243 @@
+  * If Y does not exist - X will RequeueAfter a fixed time interval
+  * If Y exists in cache - X will fetch the value to be substituted and proceed with reconciliation
+
+Currently ACK controllers do not requeue after the resource creation is successful, which means if the underlaying resource (Y) is deleted or updated those events will not get trigger any update in X. These updates will only get picked if X reconciles or user explicitly resubmits X
+ * Hence, we propose to add either a controller level or resource level flag in generator config which will add a periodic requeue even after resource creation is successful
+
+#### Advantages:
+
+* Efficient since the sharedInformer will use a push model to maintain the cache compared to pull model where each resource manager querys the API server for the resource
+* Allows submitting multiple CRs simultaneously without waiting for ARN or id to pre-exist. Name is pre-determined user input
+* In addition to supporting references, periodic requeue after resource creation can be used for
+  * Drift detection
+  * Detect destructive operations [#82](https://github.com/aws/aws-controllers-k8s/issues/82)
+
+#### Limitation (to be solved):
+
+Resource exists in cache/etcd not a sufficient condition if X refers to a field in Y.Spec. E.g.
+  * Endpoint resource (X) refers to Spec.ModelName from resource (Y). User submitted X & Y to cluster. Y.Spec has modelName when the CR is submitted but resource does not get created in AWS because of error. Since X is only looking for is Y to exist in shared informer cache, X will run create API by looking at Y.Spec.ModelName & will go into an exponential back off instead of periodic requeue
+  * If X had a dependency on Y's ARN/id (returned by AWS) then it would have requeueAfter() until Y.Status.*.ARN != nil
+
+Although, we do not want to complicate things by mixing the state of the referred-to resource with the state of the referencing resource, if the reference is created from a spec field, it needs to be treated like ARN/id. Is there a better way to solve this ?
+
+#### Exploration
+Taking deployment in K8s as an example to see behaviour
+
+* ConfigMap in Deployment:
+    * This is a referenced resource. [Updating configmap does not update the deployment/restart the pods](https://github.com/kubernetes/kubernetes/issues/22368). Doesnt look like there will be a fix as well (issue opened 4 years ago). There are [custom controllers](https://github.com/stakater/Reloader) to deal with this issue.
+
+* ReplicaSet in Deployment:
+    * This is not a referenced resource but controller watches its resources. If you delete a pod from the replicaset, it gets recreated. This cannot happen unless controller is continuously watching the pods.
+    * ```
+      k create deploy -—image nginx nginx
+      k scale deploy nginx -—replicas=3
+      k get pods
+      k delete pod <>
+      ```
```

Comment:
In the case of the ConfigMap referred to in a Deployment being changed and the Deployment controller not taking any action, that is a tricky issue indeed because there is a camp of people that *want* the Deployment controller to terminate the Pods that refer to the old ConfigMap value (and have the Deployment controller naturally spawn replacement Pods that will have the latest version of the ConfigMap).

And there is a camp of people that do *not want* the Deployment controller to do that and instead want ConfigMaps to be content-addressable and immutable -- in other words, if a Pod is launched by a Deployment that refers (via the Pod template) to ConfigMap with content at version X, then all Pods in that Deployment controller should be launched referring to that specific content in ConfigMap version X. Only if the Deployment's Pod template ConfigMap version changes should the deployment controller delete Pods and have them reload with the newly-referred ConfigMap version.

I would prefer that this proposal focus mostly on the *structure* of the AWSResourceReference object and walk through whether the fields contained in an AWSResourceReference will be sufficient for a controller to actuate some behaviour based on changes to the desired state of a referring resource or the observed state of a referred-to resource. Let's not get too in the weeds about implementation details about how controllers should/would respond to those changes.